lib/libprocstat/libprocstat.c

```diff
@@ -399,6 +405,86 @@ procstat_getc18n(struct procstat *procstat, struct kinfo_proc *kp,
 	return (-1);
 }
 
+int
+procstat_getcompartments(struct procstat *procstat, struct kinfo_proc *kp,
+    struct cheri_c18n_compart **comparts, size_t *ncompartsp)
+{
+	int name[4];
+	char *inbuf;
+	size_t n, i, cur, size;
+	struct cheri_c18n_compart *outbuf;
+
+	if (comparts == NULL || ncompartsp == NULL)
+		goto out;
+
+	switch (procstat->type) {
+	case PROCSTAT_KVM:
+		warnx("kvm method is not supported");
+		goto out;
+
+	case PROCSTAT_SYSCTL:
+		break;
+
+	case PROCSTAT_CORE:
+		warnx("core method is not supported");
+		goto out;
+
+	default:
+		warnx("unknown access method: %d", procstat->type);
+		goto out;
+	}
+
+	name[0] = CTL_KERN;
+	name[1] = KERN_PROC;
+	name[2] = KERN_PROC_C18N_COMPARTS;
+	name[3] = kp->ki_pid;
+
+	/* Estimate the size of the input buffer. */
+	if (sysctl(name, nitems(name), NULL, &n, NULL, 0) != 0) {
```

Comment:
If you adjust the sysctl estimation path as I suggested, you can just pass `&size` here (which is the more normal pattern).  It is true though that the pattern in other routines (e.g. `kinfo_getfile()`) is to go ahead and fetch the data from the sysctl into the allocated `inbuf`, then afterwards walk the `inbuf` to count the number of records to compute the `n` needed to allocate `outbuf`.  (lib/libutil/kinfo_getfile.c is an example of this pattern)